lib/mayaUsd/ufe/UsdTransform3dPointInstance.cpp

```diff
@@ -0,0 +1,270 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include "UsdTransform3dPointInstance.h"
+
+#include <mayaUsd/ufe/UsdPointInstanceOrientationModifier.h>
+#include <mayaUsd/ufe/UsdPointInstancePositionModifier.h>
+#include <mayaUsd/ufe/UsdPointInstanceScaleModifier.h>
+#include <mayaUsd/ufe/UsdPointInstanceUndoableCommands.h>
+#include <mayaUsd/ufe/UsdSceneItem.h>
+#include <mayaUsd/ufe/UsdTransform3dBase.h>
+#include <mayaUsd/ufe/UsdUndoableCommand.h>
+#include <mayaUsd/ufe/Utils.h>
+
+#include <pxr/base/gf/matrix4d.h>
+#include <pxr/base/gf/quatd.h>
+#include <pxr/base/gf/quath.h>
+#include <pxr/base/gf/rotation.h>
+#include <pxr/base/gf/transform.h>
+#include <pxr/base/gf/vec3d.h>
+#include <pxr/base/gf/vec3f.h>
+#include <pxr/base/vt/types.h>
+#include <pxr/usd/usd/attribute.h>
+#include <pxr/usd/usd/prim.h>
+#include <pxr/usd/usd/timeCode.h>
+
+#include <ufe/sceneItem.h>
+#include <ufe/transform3d.h>
+#include <ufe/transform3dHandler.h>
+#include <ufe/transform3dUndoableCommands.h>
+#include <ufe/types.h>
+
+#include <memory>
+
+PXR_NAMESPACE_USING_DIRECTIVE
+
+namespace {
+
+using namespace MayaUsd;
+using namespace MayaUsd::ufe;
+
+// Class for setMatrixCmd() implementation.
+class UsdPointInstanceSetMatrix4dUndoableCmd
+    : public UsdUndoableCommand<Ufe::SetMatrix4dUndoableCommand>
+{
+public:
+    UsdPointInstanceSetMatrix4dUndoableCmd(const Ufe::Path& path, const Ufe::Matrix4d& newMatrix)
+        : UsdUndoableCommand<Ufe::SetMatrix4dUndoableCommand>(path)
+    {
+        GfMatrix4d matrix = toUsd(newMatrix);
+        matrix.Orthonormalize();
+        const GfTransform transform(matrix);
+        _position = GfVec3f(transform.GetTranslation());
+        _orientation = GfQuath(transform.GetRotation().GetQuat());
+        _scale = GfVec3f(transform.GetScale());
+    }
+
+    ~UsdPointInstanceSetMatrix4dUndoableCmd() override = default;
+
+    bool set(const Ufe::Matrix4d&) override
+    {
+        // No-op: Maya does not set matrices through interactive manipulation.
+        TF_WARN("Illegal call to UsdPointInstanceSetMatrix4dUndoableCmd::set()");
+        return true;
+    }
+
+    void executeUndoBlock() override
+    {
+        UsdSceneItem::Ptr usdItem = downcast(sceneItem());
+
+        if (!usdItem || !usdItem->isPointInstance()) {
+            return;
+        }
+
+        const UsdPrim& usdPrim = usdItem->prim();
+        const int      instanceIndex = usdItem->instanceIndex();
+        _positionModifier.setPrimAndInstanceIndex(usdPrim, instanceIndex);
```

Comment:
Yeah, there's a bit of wishful thinking baked in here.

I'm hoping that at some point we might have some additional API to make these single-element-of-an-array type reads and writes more efficient. One idea was that if we could make the modifiers more active observers of their attribute, the modifier could actually just directly hold an array value. In that case, the modifier would just update its own array and then use that to `Set()` the `UsdAttribute`. There would be a bunch more work needed though to make sure that the modifier's copy of the array value stays in sync with what's in USD, especially when other operations might manipulate the attribute in USD.

The other thing I was considering was whether some kind of manipulation "batching" API might be desirable in UFE at some point, where we could create a single `Transform3d` that represents the same manipulation operation applied to multiple scene items. That could be another case where the modifier holds onto a whole array, the batched `Transform3d` could make a whole bunch of updates to it, and those could then be "committed" to the `UsdAttribute` when the manipulation is done.

I'm not sure whether any of those are good reasons for it to stay the way it is now since none of that currently exists, but hopefully helps explain how I ended up here.

Reply:
Yes, I think batching the write on manipulation end is the way to go.  How we get there is unclear, so let's keep what you have for now, as we don't have any definite proposals to improve upon it.